File: 215/solution.cpp
// Optimization method
// 1. Use Selection Sort not quick sort (Enhanced version of quick sort)
// 2. Don't choose last element as the pivot, "randonly choose"

class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        return quickSelect(nums, 0, nums.size() - 1, nums.size()-k);
    }
    
    int quickSelect(vector<int>& nums, int left, int right, int targetIdx) {
        while (true) {
            int pivot = partition(nums, left, right);
            
            if (pivot == targetIdx) {
                return nums[pivot];
            } else if (pivot < targetIdx) {
                left = pivot+1;
            } else {
                right = pivot-1;
            }
        }
        return 0;
    }
    
    int partition(vector<int>& nums, int left, int right) {
        
        // Randomized to find pivot value
        int pivotIdx = rand()%(right-left+1) + left;
        int pivot = nums[pivotIdx];
        
        // Movie pivot to last element, it's helpful when conducting the below logics.
        swap(nums[pivotIdx], nums[right]);
        
        // Put element less than pivot to correct place
        int storeIdx = left;
        for (int idx=left; idx < right; idx++) {
            if (nums[idx] < pivot) {
                swap(nums[storeIdx++], nums[idx]);
            }
        }
        
        // Move pivot back correct position.
        swap(nums[right], nums[storeIdx]);
        return storeIdx;
    }
};

File: 215/solution.go
import "math/rand"

func findKthLargest(nums []int, k int) int {
	return quickSelect(nums, 0, len(nums)-1, len(nums)-k)
}

func quickSelect(nums []int, left int, right int, targetIdx int) int {
	for {
		pivot := partition(nums, left, right)

		if targetIdx == pivot {
			return nums[pivot]
		} else if pivot < targetIdx {
			left = pivot + 1
		} else {
			right = pivot - 1
		}
	}
}

func partition(nums []int, left int, right int) int {
	pivotIdx := rand.Intn(right-left+1) + left
	pivot := nums[pivotIdx]

	nums[pivotIdx], nums[right] = nums[right], nums[pivotIdx]

	storeIdx := left

	for idx := left; idx < right; idx++ {
		if nums[idx] < pivot {
			nums[idx], nums[storeIdx] = nums[storeIdx], nums[idx]
			storeIdx++
		}
	}

	nums[right], nums[storeIdx] = nums[storeIdx], nums[right]
	return storeIdx
}